src/storage/txn/scheduler.rs

```diff
@@ -898,24 +879,103 @@ impl<E: Engine, L: LockManager> Scheduler<E, L> {
-        let txn_ext = snapshot.ext().get_txn_ext().cloned();
+        if self.inner.flow_controller.enabled() {
+            if self.inner.flow_controller.is_unlimited() {
+                // no need to delay if unthrottled, just call consume to record write flow
+                let _ = self.inner.flow_controller.consume(write_size);
+            } else {
+                let start = Instant::now_coarse();
+                // Control mutex is used to ensure there is only one request consuming the quota.
+                // The delay may exceed 1s, and the speed limit is changed every second.
+                // If the speed of next second is larger than the one of first second,
+                // without the mutex, the write flow can't throttled strictly.
+                let control_mutex = self.inner.control_mutex.clone();
+                let _guard = control_mutex.lock().await;
+                let delay = self.inner.flow_controller.consume(write_size);
+                let delay_end = Instant::now_coarse() + delay;
+                while !self.inner.flow_controller.is_unlimited() {
+                    let now = Instant::now_coarse();
+                    if now >= delay_end {
+                        break;
+                    }
+                    if now >= deadline.inner() {
+                        scheduler.finish_with_err(cid, StorageErrorInner::DeadlineExceeded);
+                        self.inner.flow_controller.unconsume(write_size);
+                        SCHED_THROTTLE_TIME.observe(start.saturating_elapsed_secs());
+                        return;
+                    }
+                    GLOBAL_TIMER_HANDLE
+                        .delay(std::time::Instant::now() + Duration::from_millis(1))
+                        .compat()
+                        .await
+                        .unwrap();
+                }
+                SCHED_THROTTLE_TIME.observe(start.saturating_elapsed_secs());
+            }
+        }
+
+        // Mutations on the lock CF should overwrite the memory locks.
+        //
+        // We have to remove these locks before proposing. Otherwise the locks that are migrated
+        // to other peers may include these deleted locks unexpectedly. Here is a example:
+        // 1. Here we propose deleting a lock (but not removing the lock from the memory first).
+        // 2. The region is going to transfer leader, so it proposes all its in-memory pessimistic
+        //    locks.
+        // 3. The follower deletes the lock first, and adds the migrated locks later.
+        // This leads to an incorrect result. Deleted locks appear again.
+        //
+        // Protected by the latches, be no concurrent write on the same keys. So, removing these
+        // locks now is safe. If the write fails, we will later recover these deleted locks
+        // before releasing the latches. We will not mistakenly lose the pessimistic locks.
```

Comment:
I think it is safe if the in-memory locks are removed.

If some in-memory locks are removed before this point for any reason, such as transfer leader or splitting, we can guarantee (1) the write request will not succeed because of changed term or epoch (2) the removed locks will not be added back to the new lock table because we will check term and epoch before we add them back.

By the way, I changed in https://github.com/tikv/tikv/pull/11591 to let the pessimistic locks save the term and epoch version of the region, helping us comparing them more precisely.

Reply:
Agree, no safety issues, but I think it's possible some memory locks are lost in some cases.